codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -647,22 +649,31 @@ public void onHeadersRead(ChannelHandlerContext ctx, int streamId,
         @Override
         public void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers headers,
                                   int padding, boolean endOfStream) {
-            onHttp2Frame(ctx, new DefaultHttp2HeadersFrame(headers, endOfStream, padding)
-                    .stream(requireStream(streamId)));
+            Http2FrameStream stream = requireStream(streamId);
+            onHttp2Frame(ctx, new DefaultHttp2HeadersFrame(headers, endOfStream, padding).stream(stream));
         }
 
         @Override
         public int onDataRead(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding,
                               boolean endOfStream) {
-            onHttp2Frame(ctx, new DefaultHttp2DataFrame(data, endOfStream, padding)
-                    .stream(requireStream(streamId)).retain());
+            Http2FrameStream stream = requireStream(streamId);
+            final Http2DataFrame dataframe;
+            try {
+                dataframe = new DefaultHttp2DataFrame(data.retain(), endOfStream, padding);
+            } catch (IllegalArgumentException e) {
+                // Might be thrown in case of invalid padding / length.
+                data.release();
+                throw e;
+            }
+            dataframe.stream(stream);
+            onHttp2Frame(ctx, dataframe);
             // We return the bytes in consumeBytes() once the stream channel consumed the bytes.
             return 0;
         }
 
         @Override
         public void onGoAwayRead(ChannelHandlerContext ctx, int lastStreamId, long errorCode, ByteBuf debugData) {
-            onHttp2Frame(ctx, new DefaultHttp2GoAwayFrame(lastStreamId, errorCode, debugData).retain());
+            onHttp2Frame(ctx, new DefaultHttp2GoAwayFrame(lastStreamId, errorCode, debugData.retain()));
```

Comment:
As noted above, I think the pattern of building the frame and then once we're ready to forward it calling `.retain()` was better as it didn't retain until everything else was done. With this change if the constructor throws (perhaps a constructor assert added in the future) we leak the `debugData` buffer.

Reply:
I disagree because it is kind of a miss-use of the API. We don't want to retain the frame we really want to retain the payload. And we need to do this before calling the constructor as in theory it might call release on it.

Reply:
It sounds like when we pass the content into the frame constructor we're delegating ownership of it, otherwise the constructor has no right to release the buffer etc, very similar to the notion of move in C++ and Rust. That makes a lot of sense to me.
Maybe we should make the `DefaultHttp2DataFrame` constructor (as far as I can tell it's the only one that may throw right now) release the content if it's going to throw. That would remove the odd noise in `onDataRead(..)`.